### ISSUE TITLE
Registration order of ContravariantRegistrationSource affects results

### ISSUE THREAD

**Issue by alexmg** [autofac/Autofac#341]
_From [steven.v...@gmail.com](https://code.google.com/u/100547990705934745156/) on August 27, 2011 03:09:35_

The order in wich the ContravariantRegistrationSource is registered in the container changes the behavior of the container, which could lead to hard to find problems.

The attached .cs file contains two unit tests, which only differ in the moment of registering the ContravariantRegistrationSource.

While the container in the first test returns an MultipleDispatchEventHandler<CustomerMovedEvent> when a IEventHandler<CustomerMovedEvent> is requested, the second test returns a MultipleDispatchEventHandler<object> in that case.

Note: running under .NET 4.0.

**Attachment:** [FailingMSTest.cs](http://code.google.com/p/autofac/issues/detail?id=341)

_Original issue: http://code.google.com/p/autofac/issues/detail?id=341_

```csharp
namespace TestProject2
{
    using System.Collections.Generic;
    using System.Linq;

    using Autofac;
    using Autofac.Core;
    using Autofac.Features.Variance;

    using Microsoft.VisualStudio.TestTools.UnitTesting;

    public interface IEventHandler<in TEvent>
    {
        void Handle(TEvent e);
    }

    public class CustomerMovedEvent
    {
    }

    public class CustomerMovedAbroadEvent : CustomerMovedEvent
    {
    }

    public class SpecialCustomerMovedEvent : CustomerMovedEvent
    {
    }

    public class CustomerMovedEventHandler : IEventHandler<CustomerMovedEvent>
    {
        public void Handle(CustomerMovedEvent e)
        {
        }
    }

    public class NotifyStaffWhenCustomerMovedEventHandler : IEventHandler<CustomerMovedEvent>
    {
        public void Handle(CustomerMovedEvent e)
        {
        }
    }

    public class CustomerMovedAbroadEventHandler : IEventHandler<CustomerMovedAbroadEvent>
    {
        public void Handle(CustomerMovedAbroadEvent e)
        {
        }
    }

    // A composite event handler wrapping possibly multiple handlers.
    public class MultipleDispatchEventHandler<TEvent>
        : IEventHandler<TEvent>
    {
        private IEnumerable<IEventHandler<TEvent>> handlers;

        public MultipleDispatchEventHandler(IEnumerable<IEventHandler<TEvent>> handlers)
        {
            this.handlers = handlers;
        }

        public void Handle(TEvent e)
        {
            foreach (var handler in this.handlers)
            {
                handler.Handle(e);
            }
        }
    }

    [TestClass]
    public class ContravariantRegistrationSourceDoesStrangsThingsTests
    {
        [TestMethod]
        public void SucceedingTest()
        {
            // Arrange
            var builder = new ContainerBuilder();

            // Register ContravariantRegistrationSource first
            builder.RegisterSource(new ContravariantRegistrationSource());

            builder.RegisterAssemblyTypes(typeof(IEventHandler<>).Assembly)
                .As(t => t.GetInterfaces()
                    .Where(i => i.IsClosedTypeOf(typeof(IEventHandler<>)))
                    .Select(i => new KeyedService("handler", i)));

            builder.RegisterGeneric(typeof(MultipleDispatchEventHandler<>))
                .As(typeof(IEventHandler<>))
                .WithParameter(
                    (pi, c) => pi.Name == "handlers",
                    (pi, c) => c.ResolveService(new KeyedService("handler", pi.ParameterType)));

            IContainer container = builder.Build();

            // Act
            var handler = container.Resolve<IEventHandler<CustomerMovedEvent>>();

            // Assert
            Assert.IsInstanceOfType(handler, typeof(MultipleDispatchEventHandler<CustomerMovedEvent>));
        }

        [TestMethod]
        public void FailingTest()
        {
            // Arrange
            var builder = new ContainerBuilder();

            builder.RegisterAssemblyTypes(typeof(IEventHandler<>).Assembly)
                .As(t => t.GetInterfaces()
                    .Where(i => i.IsClosedTypeOf(typeof(IEventHandler<>)))
                    .Select(i => new KeyedService("handler", i)));

            builder.RegisterGeneric(typeof(MultipleDispatchEventHandler<>))
                .As(typeof(IEventHandler<>))
                .WithParameter(
                    (pi, c) => pi.Name == "handlers",
                    (pi, c) => c.ResolveService(new KeyedService("handler", pi.ParameterType)));

            // Register ContravariantRegistrationSource last
            builder.RegisterSource(new ContravariantRegistrationSource());

            IContainer container = builder.Build();

            // Act
            var handler = container.Resolve<IEventHandler<CustomerMovedEvent>>();

            // Assert
            Assert.IsInstanceOfType(handler, typeof(MultipleDispatchEventHandler<CustomerMovedEvent>));
        }
    }
}
```

**Comment by dammejed** [comment]
Hi folks,

I got bit by this bug in a strange way just today-- some resolves which were making use of IIndex were throwing cast exceptions after I added the ContravariantRegistrationSource later in my pipeline than when I had made the .Keyed() registrations for which the IIndex was to be used. All was well once I moved the ContravariantRegistrationSource to appear before the .Keyed() registrations, however.

I was quite baffled, as the documentation only implies changes to registrations after the inclusion of the ContravariantRegistrationSource, not before. I was about to tear my hair out before I saw this issue.

Has there been any thought given to working on the issue in the near future?

**Comment by tillig** [comment]
As you can see, it's been around for a while and probably isn't super straightforward to solve. Other than what you see here, there really hasn't been any progress.

If you think you have a solution, we'd be open to a PR for it.

**Comment by dammejed** [comment]
I was afraid of that. :)

I'm certainly open to giving it a shot, though I don't have much of an idea where to start. I've been _using_ autofac with success for a while now, but I haven't delved much into the internals. 

If you had some pointers as to what avenues might lead to some success, I'd be happy to explore them.

I meant to ask in my initial comment as well: What side effects might there be when including the ContravariantRegistrationSource first as a workaround? I detected no ill effects when using IIndex<> after having included this, but IIndex is indeed a contravariant generic type, so maybe there's some subtlety in the resolution that has changed in a way I haven't easily detected.

Thanks for your prompt reply!

**Comment by tillig** [comment]
I don't really have any guidance on where to start. Set up a failing unit test and a bunch of breakpoints? This stuff can be challenging to even figure out what to look at, which is why we haven't just solved it.

No real guidance on your proposed workaround, either. Sorry. Test, test, test.

**Comment by dammejed** [comment]
OK, message received. :)

Thanks anyway for your suggestions

**Comment by tillig** [comment]
I verified that this is still a problem with the new pipeline mechanism we have and updated internals. I didn't think it would necessarily magically go away, but... well, there was always the possibility something in there got changed up such that this would just work.

I still don't see a great way to solve this, either. Registration sources get queried - just like registrations themselves - in the order in which they're registered, and the way open generics work, [that's _also_ a registration source](https://github.com/autofac/Autofac/blob/develop/src/Autofac/Features/OpenGenerics/OpenGenericRegistrationSource.cs).

We _could_ potentially have some way to override order of registration sources when they get added, similar to the `PreserveExistingDefaults` thing we have for registrations. (I'm not suggesting we actually do this, as you'll see in a moment, but we _could_ do this.)

Right now when you call [`ContainerBuilder.RegisterSource`](https://github.com/autofac/Autofac/blob/be2a437ca048a7534a5ed7936334d1336291aa1d/src/Autofac/SourceRegistrationExtensions.cs#L23) it creates a `SourceRegistrar` and calls into [`SourceRegistrar.RegisterSource`](https://github.com/autofac/Autofac/blob/be2a437ca048a7534a5ed7936334d1336291aa1d/src/Autofac/Core/Registration/SourceRegistrar.cs#L44), which is where it adds the build callback to attach the source to the container. We could add something to allow specifying _where_ in the list of sources this new one will show up.

```c#
builder.RegisterSource<T>(Position.First);
```

Such that there's a parameter that can be expressed to indicate where in the list of sources it should be added.

However, doing something like that comes with some drawbacks:

- **It's a very limited use case.** That change would basically be there _just_ to support this problem, which feels strange.
- **It still has registration order problems.** If you register _two things_ to show up in the list of sources "first," the _last one registered_ will actually be the first. Hence, order _still_ matters.

Basically, **it seems like this would fix it, but it wouldn't really fix it.**

I think the best we can do is to update the documentation on `ContravariantRegistrationSource` and indicate that if you're going to use it, it's recommended to be registered _first_ or you'll have trouble.

**Comment by tillig** [comment]
[The docs have been updated to caution folks to register this first.](https://autofac.readthedocs.io/en/latest/advanced/registration-sources.html#contravariant-registration-source) If someone figures out a great way to fix this, we can open a new issue with a proposal and, ideally, a PR at that time.
